src/runtime/pickle_runner.ts

```diff
@@ -395,9 +430,67 @@ export default class PickleRunner {
         },
       })
     }
-    return await this.invokeStep(
-      testStep.pickleStep,
-      testStep.stepDefinitions[0]
+
+    this.stepResult = undefined
+    const beforeStepHooksResult = await this.runStepHooks(
+      this.getBeforeStepHookDefinitions()
     )
+    let cumulatedStepResult = beforeStepHooksResult
+
+    if (beforeStepHooksResult.status !== Status.FAILED) {
+      const stepResult = await this.invokeStep(
+        testStep.pickleStep,
+        testStep.stepDefinitions[0]
+      )
+      if (stepResult !== undefined) {
+        this.stepResult = stepResult
+        cumulatedStepResult = stepResult
+        cumulatedStepResult.duration = addDurations(
+          cumulatedStepResult.duration,
+          beforeStepHooksResult.duration
+        )
+        const afterStepHooksResult = await this.runStepHooks(
+          this.getAfterStepHookDefinitions()
+        )
+        cumulatedStepResult.duration = addDurations(
+          cumulatedStepResult.duration,
+          afterStepHooksResult.duration
+        )
+      }
+    }
+    return cumulatedStepResult
+  }
+
+  async runStepHooks(
+    stepHooks: TestStepHookDefinition[]
+  ): Promise<messages.TestStepFinished.ITestStepResult> {
+    const stepHooksResult = messages.TestStepFinished.TestStepResult.fromObject(
+      {
+        status:
+          this.result.status === Status.FAILED
+            ? Status.SKIPPED
+            : this.result.status,
+        duration:
```

Comment:
I think this should be `getZeroDuration()` (never this.result.duration). The `this.result` duration is the sum of all steps. We want to start with zero so we only sum up the duration of this set of step hooks.